### ISSUE TITLE
ImportError: cannot import name 'secure_filename' from 'werkzeug'

### ISSUE THREAD

**Issue by kevincon** [bigeyex/python-adminui#14]
Hello, thanks for this great package! I recently tried updating to version 0.3.7 but encountered this error as soon as my project tried to `import adminui`:

```python
  File "<elided>", line <elided>, in <module>
    import adminui
  File "<elided>\lib\site-packages\adminui\__init__.py", line 1, in <module>
    from .form import *
  File "<elided>\lib\site-packages\adminui\form.py", line 2, in <module>
    from .app import callbackRegistry
  File "<elided>\lib\site-packages\adminui\app.py", line 10, in <module>
    from werkzeug import secure_filename
ImportError: cannot import name 'secure_filename' from 'werkzeug' (<elided>\lib\site-packag
es\werkzeug\__init__.py)
```

I think the issue is that `adminui` is importing from Werkzeug based on [the assumption that its Flask dependency will pull in Werkzeug](https://github.com/pallets/flask/blob/de464c03e134127140e5622e230790806a133ff9/setup.py#L13) and thus relying on Flask constraining the version of Werkzeug:

```
λ poetry show --tree adminui
adminui 0.3.7 Write professional web interface with Python
|-- flask *
|   |-- click >=5.1
|   |-- itsdangerous >=0.24
|   |-- jinja2 >=2.10.1
|   |   `-- markupsafe >=0.23
|   `-- werkzeug >=0.15
`-- pyjwt *
```

Meanwhile, the latest version of Werkzeug has moved `secure_filename` from being importable via `werkzeug` directly to now only being importable via `werkzeug.utils`. It looks like another project had to similarly update their import of `secure_filename` when they updated Werkzeug: https://github.com/maxcountryman/flask-uploads/pull/28/files#diff-a286a1482c81850709d13630b6822417R27

For now I've pinned my project's dependency on `adminui` to version 0.3.6 which does not have the new code that tries to import `secure_filename` from `werkzeug`.

Since `adminui` directly imports code from `werkzeug`, what do you think about adding `Werkzeug` to this project's list of dependencies (with an appropriate version constraint to ensure `adminui` gets the version of Werkzeug it expects)?

**Comment by Youkii-Chen** [comment]
+1

**Comment by Youkii-Chen** [comment]
see here: https://github.com/maxcountryman/flask-uploads/commit/bb79c1def3a0940f929bc2cc0038047b36b9b2d8

**Comment by Youkii-Chen** [comment]
#20

**Comment by kevincon** [comment]
@louis-shon or @Youkii-Chen, can one of you update your PR to add werkzeug as an explicit dependency of this package to help prevent this kind of issue from happening again? I believe that can be done by adding `werkzeug` to this list: https://github.com/bigeyex/python-adminui/blob/9910ced906d52c2e7cddc5ea9ba3bb835dbe3ebd/python/setup.py#L21-L24

But I would suggest adding version constraints for both `Flask` and `werkzeug` there too since `python-adminui` may not work with newer versions of either library, e.g.:

```
'Flask~=1.1.2',
'werkzeug~=0.15.6',
```

**Comment by Youkii-Chen** [comment]
i doesn't think so

`from werkzeug.utils import secure_filename` is a safety implement for all `wk` version, if another lib need wk in a same time, it may be raise a conflict

**Comment by kevincon** [comment]
I'm not sure I understand your concern, but I see that there are two risks at the moment which I believe may cause problems for users of `python-adminui` in the future:

1. `python-adminui` directly imports code from Werkzeug but does not declare that it depends on Werkzeug in the list of packages it requires in `setup.py`; instead, there is an assumption that Flask provides Werkzeug, but that may not be true in the future. If Flask ever removes its dependency on Werkzeug, then `python-adminui` will break for new users who install `python-admnui` because they will no longer automatically get Werkzeug when they `pip install python-adminui` and thus `python-adminui` will fail to import from Werkzeug at runtime. Such users could separately install Werkzeug themselves, but they may not understand that they need to do that. Instead, `python-adminui` should declare that it requires Werkzeug because it does (because it imports code from it).
2. `python-adminui` does not declare which versions of its dependencies it is compatible with in its `setup.py` file. So for example, if a new version of Flask, Werkzeug, or PyJWT is released that renames a function, module, etc. which `python-adminui` calls using the old name, then `python-adminui` will break for new users because they will get the new versions of Flask, Werkzeug, and PyJWT which would no longer be compatible with `python-adminui`. At the moment, `python-adminui` effectively declares it is compatible with all versions of Werkzeug, Flask, and PyJWT, which this issue (#14) already proves can be incorrect.

I believe adding `Werkzeug` to the `install_requires` list in `setup.py` will remove the risk of (1) and adding version constraints to each of the items in `install_requires` in `setup.py` will remove the risk of (2).

You mentioned:

>  if another lib need wk in a same time, it may be raise a conflict

But Werkzeug, as well as Flask and PyJWT, appear to use [semantic versioning](https://semver.org/), so I think it should be safe for `python-adminui` to set version constraints for these dependencies which allow for multiple different versions to be used which are backwards-compatible with `python-adminui`, e.g. using the [compatible release](https://www.python.org/dev/peps/pep-0440/#compatible-release) version specifier `~=` or something similar. If another library also requires Werkzeug, then `pip` (or whatever Python package installer/manager is being used), should be able to resolve the multiple version constraints to choose a version of Werkzeug which is compatible with `python-adminui` as well as any other libraries which require Werkzeug. And if it can't, a conflict should be raised because that means `python-adminui` really may not be compatible with those other libraries.

When new versions of Werkzeug, Flask, and PyJWT are released, the version constraints in `python-adminui`'s `setup.py` file can be updated and a new version of `python-adminui` can be released after it is verified that `python-adminui` really does support those new versions.

Does that make sense? Or if you still disagree, can you provide an example to explain further to help me understand your concern?

**Comment by Youkii-Chen** [comment]
What you said makes sense, but in general, a support library will only specify the minimum version of the dependent module (>=xxx ver).

Suppose that module a stipulates that the 0.1 version of flask must be used, and then when you install module b which stipulate that you need to use the 0.2 version of flask. In this case, module b may not be installed, or it may cause module a to run incorrectly.

Therefore, we should continue to maintain this project to make it and its dependencies work well at the same time.

Therefore, we should not set the upper limit version for this project, but set the lower limit version, and it should be left to the author to determine which minimum version to use.

About (2), if it is not necessary, their developers will not do this, and it is their responsibility to make their projects have good downward compatibility.

**Comment by kevincon** [comment]
Thanks for sharing some more details to help me understand your concern a little better. I think the problem is that in either:

i. the scenario you mentioned where modules `a` and `b` are not installed together because they have incompatible version constraints for Flask
ii. the scenario of this issue (#14) where an incompatible version of Werkzeug was installed because `pyadmin-ui` does not constrain the maximum version of Flask/Werkzeug it supports

users of `python-adminui` experience a problem which requires a change to be made in `python-adminui` in order to fix the problem. 

But I believe that the problem they experience in (i) is a little nicer than the problem they experience in (ii) because in (i) I think they should see a fairly clear error message from pip (or whatever package installer they use) explaining that no version of Flask can be installed which matches the requirements of `a` and `b`, but in (ii) they see an error like this `ImportError: cannot import name 'secure_filename' from 'werkzeug'` which is less clear because it does not indicate the cause of the problem (incompatible versions) but rather a symptom (that a symbol cannot be imported).

> in general, a support library will only specify the minimum version of the dependent module (>=xxx ver).

I don't think it's as general a practice as you might think, e.g. see:

* requests: https://github.com/psf/requests/blob/590350f8d094c216051510ed1dd18fe871b53b72/setup.py#L44-L49
* scipy: https://github.com/scipy/scipy/blob/7c904a77ad3d239bc35f18222810c9454e5441c3/pyproject.toml#L1-L4
* celery: https://github.com/celery/celery/blob/ef77fcd2ac872275cdd0f85e21180fe7b6433125/requirements/default.txt#L2-L5

Will you at least address the risk I mentioned in (1) by adding Werkzeug to `install_requires` in your PR, even if you would prefer not setting a maximum version constraint for it?

**Comment by bigeyex** [comment]
Thanks for all the insightful discussions! 
For now, I merged pr#20, and added werkzeug to install_requires. 
If there's still concern about the minimal version of libraries or such, discussions are welcomed.

**Comment by arslan-ahmad-dev** [comment]
Solution : pip install Flask-Reuploaded
